rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,436 @@
+pass extra information around (otherwise it would be equivalent to
+pass layout tag / brgemm object around).  Current internal
+implementation of layout selection depends on highest available ISA,
+input datatype, fp_math mode and accumulation datatype.
+
+In general, we would recommend to tie the layout query to a brgemm
+object for two reasons:
+- from oneDNN implementation perspective, we can extend the
+  information used to select best layout
+- in practice, when data is pre-packed, blocking information (so brgemm
+  object creation parameters) have to be known in order to layout data
+  in an optimal way.
+
+In order to make the case where brgemm objects are recreated for the
+sole purpose of packing data, brgemm obect creation will be split in
+two steps (brgemm::desc and brgemm classes) to avoid jitting
+overheads.  This will introduce some overhead for brgemm objects
+creation though (potential extra copies upon brgemm constructor call).
+
+#### Other considerations
+
+The recommendation as a starting point is to make query mandatory: no
+layout forcing on user side. This simplifies the configurations
+supported by brgemm. Given that we have no support internally for
+transa/tranb, those will not be exposed as well. We will also support
+only row major representation, and not expose a knob to select
+row/column major.
+
+Given the small number of layouts, we will expose a layout tag enum
+(next section will address the lack of need for metadata handling by
+oneDNN brgemm).
+
+## Low precision handling.
+
+The API will be limited to support downconversion/upconversion before
+computation in a similar way that
+[dnnl::fpmath_mode](https://oneapi-src.github.io/oneDNN/dev_guide_attributes_fpmath_mode.html#)
+provides.
+
+If a user wants to implement quantized computation with zero-point,
+they will have to precompute compensation term and pass it to brgemm
+computation as bias.
+
+The main drivers behind this choice are:
+- flexibility on the user side, as any custom quantization scheme can
+  be handled. It removes the dependency on oneDNN wrt when/how any
+  compensation term should be computed.
+- consistency on the user side. Whether they implement a custom
+  quantization scheme or a standard one, the programming model is the
+  same.
+- supporting metadata on oneDNN sides would force using higher level
+  abstraction that can impact performance. Also, there is little to no
+  performance gain to expect from this level of fusion.
+
+## Hardware specific discussions
+
+### ISA selection
+
+Here two options: 
+- the user can specify for which ISA the kernel can be generated,
+- oneDNN does it transparently under the hood to generate best kernel
+  for current platform.
+
+Note that current oneDNN implementation just takes ISA as a maximum
+ISA.  To simplify the programming model, isa selection will not appear
+in block level APIs. However, `dnnl::set_max_cpu_isa` will still be
```

Comment:
In this context, if there is no ISA setting at the BRGEMM level, I don't think we will use this global API to configure things. I assume oneDNN would make wise decision to choose right ISA given input shapes, is that correct?

Reply:
Yes the plan is to pick best available ISA under the hood. If ISA control is needed (there are some corner cases), that global API can be used.

Reply:
Using the global API will cause problem when do parallel compiling, like one thread want to generate AVX512 brgemm while another thread want to generate AMX brgemm. Suggest to have another way to specify the ISA per brgemm creation.

Reply:
> Yes the plan is to pick best available ISA under the hood. If ISA control is needed (there are some corner cases), that global API can be used.

What do you mean by "best available ISA"? If AVX is a better choice for a particular shape than AMX, will you choose AVX or AMX?

Reply:
@ZhennanQin do you have a usage scenario for using different ISA in a single process?

@jgong5 best available means highest performing ISA on the platform that is creating that brgemm object (e.g. AMX > AVX512 > AVX2 ..., more details [here](https://oneapi-src.github.io/oneDNN/dev_guide_cpu_dispatcher_control.html))

In the current proposal, we cannot use the shape to select ISA, since ISA selection could impact data layout (e.g. we currently might not require 32-bit packed/VNNI layout when lower ISA requires emulation).
We could make layout independent of ISA but this would require changes in current oneDNN kernel generation logic (or disabling some datatype/isa combinations).

Reply:
> In the current proposal, we cannot use the shape to select ISA, since ISA selection could impact data layout (e.g. we currently might not require 32-bit packed/VNNI layout when lower ISA requires emulation).
> We could make layout independent of ISA but this would require changes in current oneDNN kernel generation logic (or disabling some datatype/isa combinations).

Suppose we want to compute with int8 (a typical case is woq int4 with int8 compute), INT8 VNNI might work better than INT8 AMX with smaller shapes while both would have same VNNI4 weight layout. How do you decide the ISA in this case?

Reply:
> @ZhennanQin do you have a usage scenario for using different ISA in a single process?

A valid usage scenario is, if the input shape is not dividable by the tiling size, we may use a small brgemm to handle the tail shape. This small brgemm only execute once, AVX512-VNNI may be faster than AMX considering the tile reconfigure overhead. Then for this mamtul, both AMX and AVX512-VNNI brgemms are used.

Reply:
@jgong5 in the case you provide, there is no implication of isa wrt. Currently, isa has layout implication mostly when f16 emulation happens. But there are other scenarios, where some architectures have different instructions for matrix multiplication acceleration that require different layout (e.g. bfmmla, and SME on ARM).

> How do you decide the ISA in this case?

We just use the highest available ISA for this datatype, independently of shape. See [here ](https://github.com/oneapi-src/oneDNN/blob/cf5a4fedb42bdf54e06269208945459c699890c2/src/cpu/x64/brgemm/brgemm_utils.cpp#L114) for the exact logic

> Then for this mamtul, both AMX and AVX512-VNNI brgemms are used.

Makes sense, but why would you create those in separate threads?

Reply:
> Makes sense, but why would you create those in separate threads?

MLIR support multi-thread compilation, we can't guarantee the brgemm creations happen within a single thread or separate threads.